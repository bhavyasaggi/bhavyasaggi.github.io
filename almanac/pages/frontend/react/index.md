---
title: React
eleventyNavigation:
  parent: /pages/frontend/
  order: 21
---

# React

## Styling Libraries

### Stylesheet Generators

- Less / Sass / Scss
- bootstrap
- tailwind
- vanilla-extract

### Styled Components

- css-modules
- styled-components
- emotion.js

## Iconography

## Component Libraries

- React-Bootstrap
- Ant Design
- MUI

## Functionality Libraries

- Formik
- Framer Motion
- RxJS
