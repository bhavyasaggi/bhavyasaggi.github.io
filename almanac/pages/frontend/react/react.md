---
title: React
eleventyNavigation:
  parent: /pages/frontend/
  order: 21
---

# React

## Styling Libraries

### Preprocessors (Less / Sass / Scss)

### Stylesheet Generators

- bootstrap
- tailwind
- vanilla-extract

### Styled Components

- css-modules
- styled-components
- emotion.js

## Iconography

## Component Libraries

- React-Bootstrap
- Ant Design
- MUI

## Functionality Libraries

- Formik
- Framer Motion
- RxJS
