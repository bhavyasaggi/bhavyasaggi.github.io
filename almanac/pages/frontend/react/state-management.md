---
title: State Management
eleventyNavigation:
  parent: /pages/frontend/react/
  order: 1
---

# State Management

## Global Store

- Redux

## Atomic State

- Recoil / jotai

- Zustand

## Query Based

- RTK / Apollo
