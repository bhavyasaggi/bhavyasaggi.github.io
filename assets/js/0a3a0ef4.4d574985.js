"use strict";(self.webpackChunkalmanac=self.webpackChunkalmanac||[]).push([[5397],{3769:a=>{a.exports=JSON.parse('{"name":"docusaurus-plugin-content-docs","id":"default"}')}}]);