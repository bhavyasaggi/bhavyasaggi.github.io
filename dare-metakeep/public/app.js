'use strict'

import React, {
  StrictMode,
  useRef,
  useState,
  useReducer,
  useCallback,
  useEffect,
} from 'react'
import ReactDOM from 'react-dom'

import {
  Container,
  Row,
  Col,
  Navbar,
  Card,
  Table,
  ToggleButtonGroup,
  ToggleButton,
  Form,
  InputGroup,
} from 'react-bootstrap'

import {
  Chart as ChartJS,
  BarElement,
  LineElement,
  PointElement,
  Title,
  Tooltip,
  Legend,
  LinearScale,
  TimeScale,
} from 'chart.js'
import { Bar, Line } from 'react-chartjs-2'
import { isValid, format, subDays } from 'date-fns'
import 'chartjs-adapter-date-fns'

ChartJS.register(
  BarElement,
  LineElement,
  PointElement,
  Title,
  Tooltip,
  Legend,
  LinearScale,
  TimeScale
)

export const QUERY_DATETIME = {
  LAST_24_HOURS: 'Last 24 Hours',
  LAST_7_DAYS: 'Last 7 days',
  CUSTOM_RANGE: 'Custom Range',
}

const isValidDate = (date) => isValid(new Date(date))

export function CardCount({ title, value, StatusComponent }) {
  return (
    <Card className='text-center'>
      <Card.Header>{title}</Card.Header>
      <Card.Body>
        {StatusComponent ? <StatusComponent /> : value || 0}
      </Card.Body>
    </Card>
  )
}

export function CardTimeseries({ title, value, StatusComponent }) {
  const dataCount = []
  const dataUsers = []
  const dataFailures = []
  ;(value || []).forEach(
    ({ timestamp, countTotal, countUsers, countFailures }) => {
      if (!isValidDate(timestamp)) {
        return
      }
      const datetimestamp = new Date(timestamp)
      dataCount.push({ x: datetimestamp, y: countTotal })
      dataUsers.push({ x: datetimestamp, y: countUsers })
      dataFailures.push({ x: datetimestamp, y: countFailures })
    }
  )

  return (
    <Card className='text-center'>
      <Card.Header>{title}</Card.Header>
      <Card.Body
        className='d-flex align-items-center justify-content-center'
        style={{
          height: '300px',
        }}
      >
        {StatusComponent ? (
          <StatusComponent />
        ) : (
          <Line
            data={{
              datasets: [
                {
                  label: 'Total Log Count',
                  data: dataCount,
                  backgroundColor: 'blue',
                  borderColor: 'blue',
                  tension: 0.1,
                },
                {
                  label: 'Distinct User Count',
                  data: dataUsers,
                  backgroundColor: 'green',
                  borderColor: 'green',
                  tension: 0.1,
                },
                {
                  label: 'Failure Log Count',
                  data: dataFailures,
                  backgroundColor: 'red',
                  borderColor: 'red',
                  tension: 0.1,
                },
              ],
            }}
            options={{
              responsive: true,
              maintainAspectRatio: false,
              scales: {
                x: {
                  type: 'time',
                  time: {
                    minUnit: 'minute',
                  },
                },
                y: {},
              },
            }}
          />
        )}
      </Card.Body>
    </Card>
  )
}

export function CardTimetable({ title, value, StatusComponent }) {
  return (
    <Card>
      <Card.Header>{title}</Card.Header>
      {StatusComponent ? (
        <Card.Body
          className='d-flex align-items-center justify-content-center'
          style={{
            minHeight: '300px',
          }}
        >
          <StatusComponent />
        </Card.Body>
      ) : (
        <Table size='sm' responsive bordered striped hover>
          <thead>
            <tr>
              <td>#</td>
              <td>UserId</td>
              <td>timestamp</td>
              <td>Status</td>
              <td>Error Message</td>
              <td>Request</td>
              <td>Response</td>
            </tr>
          </thead>
          <tbody className='small'>
            {(value || []).map(
              (
                { key, userId, timestamp, status, errorMsg, request, response },
                idx
              ) => (
                <tr key={key || idx}>
                  <td>{idx + 1}</td>
                  <td>{userId}</td>
                  <td>
                    {isValidDate(timestamp) ? (
                      <time dateTime={timestamp}>
                        {format(
                          new Date(timestamp),
                          'MMM d, yyyy, h:mm:ss.SSS aaaa'
                        )}
                      </time>
                    ) : (
                      '-'
                    )}
                  </td>
                  <td>{status === 0 ? 'Success' : 'Failure'}</td>
                  <td>{errorMsg || ''}</td>
                  <td>{JSON.stringify(request)}</td>
                  <td>{JSON.stringify(response)}</td>
                </tr>
              )
            )}
          </tbody>
        </Table>
      )}
    </Card>
  )
}

export function FilterDateTime({
  queryDateTime,
  queryDateTimeFrom,
  queryDateTimeTo,
  updateDataMeta,
}) {
  const formRef = useRef(null)

  const formNoop = useCallback((formEvent) => {
    formEvent.preventDefault()
  }, [])
  const formAction = useCallback(() => {
    if (formRef.current) {
      const formData = Object.fromEntries(
        new FormData(formRef.current).entries()
      )
      const deltaDataMeta = {
        queryDateTime: formData.dType,
      }
      if (formData.dFrom) {
        deltaDataMeta.queryDateTimeFrom = new Date(formData.dFrom).getTime()
      }
      if (formData.dTo) {
        deltaDataMeta.queryDateTimeTo = new Date(formData.dTo).getTime()
      }
      updateDataMeta(deltaDataMeta)
    }
  }, [])

  const isDisabledDateTime = queryDateTime !== QUERY_DATETIME.CUSTOM_RANGE

  return (
    <Form
      ref={formRef}
      onChange={formAction}
      onSubmit={formNoop}
      onReset={formNoop}
      className='border rounded p-3'
    >
      <Row>
        <Col>
          <Row>
            <Col>
              <InputGroup>
                <InputGroup.Text>From</InputGroup.Text>
                <Form.Control
                  name='dFrom'
                  type='datetime-local'
                  value={
                    format(queryDateTimeFrom, 'yyyy-MM-dd') +
                    'T' +
                    format(queryDateTimeFrom, 'HH:mm')
                  }
                  disabled={isDisabledDateTime}
                  data-testid='dType-datetime-start'
                />
              </InputGroup>
            </Col>
            <Col>
              <InputGroup>
                <InputGroup.Text>To</InputGroup.Text>
                <Form.Control
                  name='dTo'
                  type='datetime-local'
                  value={
                    format(queryDateTimeTo, 'yyyy-MM-dd') +
                    'T' +
                    format(queryDateTimeTo, 'HH:mm')
                  }
                  disabled={isDisabledDateTime}
                  data-testid='dType-datetime-end'
                />
              </InputGroup>
            </Col>
          </Row>
        </Col>
        <Col className='text-end'>
          <ToggleButtonGroup type='radio' name='dType' value={queryDateTime}>
            {Object.entries(QUERY_DATETIME).map(
              ([queryDateTimeKey, queryDateTimeValue]) => (
                <ToggleButton
                  key={queryDateTimeKey}
                  id={`dType-toggle-${queryDateTimeKey}`}
                  data-testid={`dType-${queryDateTimeKey}`}
                  value={queryDateTimeValue}
                >
                  {queryDateTimeValue}
                </ToggleButton>
              )
            )}
          </ToggleButtonGroup>
        </Col>
      </Row>
    </Form>
  )
}

// TODO: replace with useSWR
export function useFetch() {
  const fetchCb = useCallback((resource, options = {}) => {
    const abortController = new AbortController()
    const fetchPromise = () =>
      fetch(resource, { ...options, signal: abortController.signal })
    const fetchAbort = () => {
      abortController.abort()
    }
    return [fetchPromise, fetchAbort]
  }, [])
  return fetchCb
}

export const DEFAULT_DATA_SET = {}
export const DEFAULT_DATA_REDUCER = (prevMeta, deltaMeta) => {
  const nextMeta = { ...prevMeta, ...deltaMeta }
  if (prevMeta.queryDateTime !== nextMeta.queryDateTime) {
    const currentDateTime = new Date().getTime()
    switch (nextMeta.queryDateTime) {
      case QUERY_DATETIME.LAST_24_HOURS:
        nextMeta.queryDateTimeTo = currentDateTime
        nextMeta.queryDateTimeFrom = subDays(currentDateTime, 1).getTime()
        break
      case QUERY_DATETIME.LAST_7_DAYS:
        nextMeta.queryDateTimeTo = currentDateTime
        nextMeta.queryDateTimeFrom = subDays(currentDateTime, 7).getTime()
        break
      default:
        break
    }
  }
  if (nextMeta.queryDateTimeFrom > nextMeta.queryDateTimeTo) {
    window.alert('Mismatched TimeRange')
    return prevMeta
  }
  // TODO: deep compare
  return nextMeta
}
export const DEFAULT_DATA_META = DEFAULT_DATA_REDUCER(
  {
    isLoading: true,
    isError: false,
  },
  {
    queryDateTime: QUERY_DATETIME.LAST_24_HOURS,
    queryDateTimeFrom: new Date().getTime(),
    queryDateTimeTo: new Date().getTime(),
  }
)

export default function App() {
  // {count: {total, failures, users}, data: [...{userId, timestamp, status, errorMsg, request, response}]}
  const [dataset, setDataset] = useState(DEFAULT_DATA_SET)
  const [
    { isLoading, isError, queryDateTime, queryDateTimeFrom, queryDateTimeTo },
    updateDataMeta,
  ] = useReducer(DEFAULT_DATA_REDUCER, DEFAULT_DATA_META)

  const fetchCb = useFetch()
  useEffect(() => {
    updateDataMeta({ isLoading: true, isError: false })

    // TODO: Update with appropriate API Path
    const resource = new URL('/api/logs', window.location.origin)
    resource.searchParams.append('from', queryDateTimeFrom)
    resource.searchParams.append('to', queryDateTimeTo)

    const [fetchPromise, fetchAbort] = fetchCb(resource.toString())

    // TODO: Debounce
    fetchPromise()
      .then((res) => res.json())
      .then((data) => {
        setDataset(data)
        updateDataMeta({ isLoading: false, isError: false })
      })
      .catch(() => {
        setDataset(DEFAULT_DATA_SET)
        updateDataMeta({ isLoading: false, isError: true })
      })
    return () => {
      fetchAbort()
      updateDataMeta({ isLoading: false, isError: false })
    }
  }, [queryDateTimeFrom, queryDateTimeTo])

  let StatusComponent = null
  if (isLoading) {
    StatusComponent = () => '⏳'
  }
  if (isError) {
    StatusComponent = () => '⚠️'
  }
  if (!dataset || !dataset.data || !dataset.data.length) {
    StatusComponent = () => '⛔'
  }

  return (
    <>
      <Navbar variant='dark' bg='dark'>
        <Container>
          <Navbar.Brand as='header'>Hello World Log Visualizer</Navbar.Brand>
        </Container>
      </Navbar>
      <Container className='py-3'>
        <Row>
          <Col>
            <FilterDateTime
              queryDateTime={queryDateTime}
              queryDateTimeFrom={queryDateTimeFrom}
              queryDateTimeTo={queryDateTimeTo}
              updateDataMeta={updateDataMeta}
            />
          </Col>
        </Row>
      </Container>
      <Container as='main' className='py-3'>
        <Row className='mb-3'>
          <Col>
            <CardCount
              title='Total Logs'
              value={dataset?.count?.total || 0}
              StatusComponent={StatusComponent}
            />
          </Col>
          <Col>
            <CardCount
              title='Total Users'
              value={dataset?.count?.users || 0}
              StatusComponent={StatusComponent}
            />
          </Col>
          <Col>
            <CardCount
              title='Total Failures'
              value={dataset?.count?.failures || 0}
              StatusComponent={StatusComponent}
            />
          </Col>
        </Row>
        <Row className='mb-3'>
          <Col>
            <CardTimeseries
              title='Log Summary'
              value={dataset?.dataSummary}
              StatusComponent={StatusComponent}
            />
          </Col>
        </Row>
        <Row className='mb-3'>
          <Col>
            <CardTimetable
              title='Logs'
              value={dataset.data}
              StatusComponent={StatusComponent}
            />
          </Col>
        </Row>
      </Container>
      <footer>
        <Container>
          <Row>
            <Col className='text-center'>Copyright</Col>
          </Row>
        </Container>
      </footer>
    </>
  )
}

// To bypass in jest env
if (typeof window !== 'undefined' && window.__ENV__) {
  ReactDOM.render(
    <StrictMode>
      <App />
    </StrictMode>,
    document.getElementById('main')
  )
}
