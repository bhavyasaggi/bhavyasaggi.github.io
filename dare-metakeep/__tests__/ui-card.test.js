/**
 * @jest-environment jsdom
 */
import React from 'react'
import { render, waitFor } from '@testing-library/react'
import 'jest-canvas-mock'

import { CardCount, CardTimeseries, CardTimetable } from '../public/app.js'

const DUMMY_DATA = [
  {
    key: '6560d4c7f986db1c69532ab4',
    userId: '1',
    timestamp: '2023-11-24T16:52:23.168Z',
    request: {
      hostname: 'localhost',
      ip: '::1',
      path: '/api/hello-world',
      xhr: false,
      query: {
        userId: '1',
      },
      body: {},
    },
    response: {
      userId: '1',
      timestamp: '2023-11-24T16:52:23.168Z',
      status: 0,
    },
    status: 0,
  },
  {
    key: '6561020762eccd1e60d847df',
    userId: '1',
    timestamp: '2023-11-24T16:52:23.168Z',
    request: {
      hostname: 'localhost',
      ip: '::1',
      path: '/api/hello-world',
      xhr: false,
      query: {
        userId: '1',
      },
      body: {},
    },
    response: {
      userId: '1',
      timestamp: '2023-11-24T16:52:23.168Z',
      status: 0,
    },
    status: 1,
  },
  {
    key: '6560d4c9f986db1c69532ab6',
    userId: '2',
    timestamp: '2023-11-24T16:52:25.388Z',
    request: {
      hostname: 'localhost',
      ip: '::1',
      path: '/api/hello-world',
      xhr: false,
      query: {
        userId: '2',
      },
      body: {},
    },
    response: {
      userId: '2',
      timestamp: '2023-11-24T16:52:25.388Z',
      status: 1,
      errorMsg: 'Random Failure',
    },
    status: 1,
    errorMsg: 'Random Failure',
  },
  {
    key: '6560d4cbf986db1c69532ab8',
    userId: '3',
    timestamp: '2023-11-24T16:52:27.781Z',
    request: {
      hostname: 'localhost',
      ip: '::1',
      path: '/api/hello-world',
      xhr: false,
      query: {
        userId: '3',
      },
      body: {},
    },
    response: {
      userId: '3',
      timestamp: '2023-11-24T16:52:27.781Z',
      status: 0,
    },
    status: 0,
  },
]

const DUMMY_DATA_WRONG = JSON.parse(JSON.stringify(DUMMY_DATA)).map((datum) => {
  datum.timestamp = NaN
  return datum
})

beforeAll(() => {
  class ResizeObserver {
    observe() {}
    unobserve() {}
    disconnect() {}
  }
  window.ResizeObserver = ResizeObserver
})
afterAll(() => {
  jest.clearAllMocks()
})

describe('CardCount', () => {
  it('Empty Data', () => {
    const { asFragment } = render(
      <CardCount title='Test Empty Render' value={0} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Dummy Data', () => {
    const { asFragment } = render(
      <CardCount title='Test Dummy Render' value={3} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Malformed Data', () => {
    const { asFragment } = render(
      <CardCount title='Test Malformed Render' value={NaN} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Loading State', () => {
    const { asFragment } = render(
      <CardCount
        title='Test Loading Render'
        value={NaN}
        StatusComponent={() => 'TRL'}
      />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Error State', () => {
    const { asFragment } = render(
      <CardCount
        title='Test Error Render'
        value={NaN}
        StatusComponent={() => 'TRE'}
      />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
})

describe('CardTimeseries', () => {
  it('Empty Data', () => {
    const { asFragment } = render(
      <CardTimeseries title='Test Empty Render' value={[]} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Dummy Data', () => {
    const { asFragment } = render(
      <CardTimeseries title='Test Dummy Render' value={DUMMY_DATA} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Malformed Data', () => {
    const { asFragment } = render(
      <CardTimeseries title='Test Malformed Render' value={DUMMY_DATA_WRONG} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Loading State', () => {
    const { asFragment } = render(
      <CardTimeseries
        title='Test Loading Render'
        value={NaN}
        StatusComponent={() => 'TRL'}
      />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Error State', () => {
    const { asFragment } = render(
      <CardTimeseries
        title='Test Loading Render'
        value={NaN}
        StatusComponent={() => 'TRE'}
      />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
})

describe('CardTimetable', () => {
  it('Empty Data', () => {
    const { asFragment } = render(
      <CardTimetable title='Test Empty Render' value={[]} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Dummy Data', () => {
    const { asFragment } = render(
      <CardTimetable title='Test Dummy Render' value={DUMMY_DATA} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Malformed Data', () => {
    const { asFragment } = render(
      <CardTimetable title='Test Malformed Render' value={DUMMY_DATA_WRONG} />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Loading State', () => {
    const { asFragment } = render(
      <CardTimetable
        title='Test Loading Render'
        value={NaN}
        StatusComponent={() => 'TRL'}
      />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
  it('Error State', () => {
    const { asFragment } = render(
      <CardTimetable
        title='Test Loading Render'
        value={NaN}
        StatusComponent={() => 'TRE'}
      />
    )
    waitFor(async () => {
      await expect(asFragment()).toMatchSnapshot()
    })
  })
})
