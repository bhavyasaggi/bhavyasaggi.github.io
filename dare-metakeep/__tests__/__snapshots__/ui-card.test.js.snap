// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardCount Dummy Data 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Dummy Render
    </div>
    <div
      class="card-body"
    >
      3
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardCount Empty Data 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Empty Render
    </div>
    <div
      class="card-body"
    >
      0
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardCount Error State 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Error Render
    </div>
    <div
      class="card-body"
    >
      TRE
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardCount Loading State 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Loading Render
    </div>
    <div
      class="card-body"
    >
      TRL
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardCount Malformed Data 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Malformed Render
    </div>
    <div
      class="card-body"
    >
      0
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimeseries Dummy Data 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Dummy Render
    </div>
    <div
      class="d-flex align-items-center justify-content-center card-body"
      style="height: 300px;"
    >
      <canvas
        height="0"
        role="img"
        style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
        width="0"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimeseries Empty Data 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Empty Render
    </div>
    <div
      class="d-flex align-items-center justify-content-center card-body"
      style="height: 300px;"
    >
      <canvas
        height="0"
        role="img"
        style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
        width="0"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimeseries Error State 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Loading Render
    </div>
    <div
      class="d-flex align-items-center justify-content-center card-body"
      style="height: 300px;"
    >
      TRE
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimeseries Loading State 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Loading Render
    </div>
    <div
      class="d-flex align-items-center justify-content-center card-body"
      style="height: 300px;"
    >
      TRL
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimeseries Malformed Data 1`] = `
<DocumentFragment>
  <div
    class="text-center card"
  >
    <div
      class="card-header"
    >
      Test Malformed Render
    </div>
    <div
      class="d-flex align-items-center justify-content-center card-body"
      style="height: 300px;"
    >
      <canvas
        height="0"
        role="img"
        style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
        width="0"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimetable Dummy Data 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Test Dummy Render
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-sm table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <td>
              #
            </td>
            <td>
              UserId
            </td>
            <td>
              timestamp
            </td>
            <td>
              Status
            </td>
            <td>
              Error Message
            </td>
            <td>
              Request
            </td>
            <td>
              Response
            </td>
          </tr>
        </thead>
        <tbody
          class="small"
        >
          <tr>
            <td>
              1
            </td>
            <td>
              1
            </td>
            <td>
              <time
                datetime="2023-11-24T16:52:23.168Z"
              >
                Nov 24, 2023, 10:22:23.168 p.m.
              </time>
            </td>
            <td>
              Success
            </td>
            <td />
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"1"},"body":{}}
            </td>
            <td>
              {"userId":"1","timestamp":"2023-11-24T16:52:23.168Z","status":0}
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              1
            </td>
            <td>
              <time
                datetime="2023-11-24T16:52:23.168Z"
              >
                Nov 24, 2023, 10:22:23.168 p.m.
              </time>
            </td>
            <td>
              Failure
            </td>
            <td />
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"1"},"body":{}}
            </td>
            <td>
              {"userId":"1","timestamp":"2023-11-24T16:52:23.168Z","status":0}
            </td>
          </tr>
          <tr>
            <td>
              3
            </td>
            <td>
              2
            </td>
            <td>
              <time
                datetime="2023-11-24T16:52:25.388Z"
              >
                Nov 24, 2023, 10:22:25.388 p.m.
              </time>
            </td>
            <td>
              Failure
            </td>
            <td>
              Random Failure
            </td>
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"2"},"body":{}}
            </td>
            <td>
              {"userId":"2","timestamp":"2023-11-24T16:52:25.388Z","status":1,"errorMsg":"Random Failure"}
            </td>
          </tr>
          <tr>
            <td>
              4
            </td>
            <td>
              3
            </td>
            <td>
              <time
                datetime="2023-11-24T16:52:27.781Z"
              >
                Nov 24, 2023, 10:22:27.781 p.m.
              </time>
            </td>
            <td>
              Success
            </td>
            <td />
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"3"},"body":{}}
            </td>
            <td>
              {"userId":"3","timestamp":"2023-11-24T16:52:27.781Z","status":0}
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimetable Empty Data 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Test Empty Render
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-sm table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <td>
              #
            </td>
            <td>
              UserId
            </td>
            <td>
              timestamp
            </td>
            <td>
              Status
            </td>
            <td>
              Error Message
            </td>
            <td>
              Request
            </td>
            <td>
              Response
            </td>
          </tr>
        </thead>
        <tbody
          class="small"
        />
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimetable Error State 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Test Loading Render
    </div>
    <div
      class="d-flex align-items-center justify-content-center card-body"
      style="min-height: 300px;"
    >
      TRE
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimetable Loading State 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Test Loading Render
    </div>
    <div
      class="d-flex align-items-center justify-content-center card-body"
      style="min-height: 300px;"
    >
      TRL
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardTimetable Malformed Data 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Test Malformed Render
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-sm table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <td>
              #
            </td>
            <td>
              UserId
            </td>
            <td>
              timestamp
            </td>
            <td>
              Status
            </td>
            <td>
              Error Message
            </td>
            <td>
              Request
            </td>
            <td>
              Response
            </td>
          </tr>
        </thead>
        <tbody
          class="small"
        >
          <tr>
            <td>
              1
            </td>
            <td>
              1
            </td>
            <td>
              -
            </td>
            <td>
              Success
            </td>
            <td />
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"1"},"body":{}}
            </td>
            <td>
              {"userId":"1","timestamp":"2023-11-24T16:52:23.168Z","status":0}
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              1
            </td>
            <td>
              -
            </td>
            <td>
              Failure
            </td>
            <td />
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"1"},"body":{}}
            </td>
            <td>
              {"userId":"1","timestamp":"2023-11-24T16:52:23.168Z","status":0}
            </td>
          </tr>
          <tr>
            <td>
              3
            </td>
            <td>
              2
            </td>
            <td>
              -
            </td>
            <td>
              Failure
            </td>
            <td>
              Random Failure
            </td>
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"2"},"body":{}}
            </td>
            <td>
              {"userId":"2","timestamp":"2023-11-24T16:52:25.388Z","status":1,"errorMsg":"Random Failure"}
            </td>
          </tr>
          <tr>
            <td>
              4
            </td>
            <td>
              3
            </td>
            <td>
              -
            </td>
            <td>
              Success
            </td>
            <td />
            <td>
              {"hostname":"localhost","ip":"::1","path":"/api/hello-world","xhr":false,"query":{"userId":"3"},"body":{}}
            </td>
            <td>
              {"userId":"3","timestamp":"2023-11-24T16:52:27.781Z","status":0}
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
